homeassistant/components/drop_connect/coordinator.py

```diff
@@ -0,0 +1,202 @@
+from homeassistant.util.json import JSON_DECODE_EXCEPTIONS, json_loads
+
+from .const import (
+    CONF_DATA_TOPIC,
+    CONF_DEVICE_DESC,
+    CONF_DEVICE_ID,
+    CONF_DEVICE_TYPE,
+    CONF_HUB_ID,
+    DEV_HUB,
+    DOMAIN,
+    KEY_STATUS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DROPDeviceDataUpdateCoordinator(DataUpdateCoordinator):
+    """DROP device object."""
+
+    config_entry: ConfigEntry
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Initialize the device."""
+        self._model: str = f"{config_entry.data[CONF_DEVICE_DESC]} on hub {config_entry.data[CONF_HUB_ID]}"
+        if config_entry.data[CONF_DEVICE_TYPE] == DEV_HUB:
+            self._model = f"Hub {config_entry.data[CONF_HUB_ID]}"
+        self._manufacturer: str = "Chandler Systems, Inc."
+        self._device_name: str = config_entry.data["name"]
+        self._device_information: dict[str, Any] = {}
+        super().__init__(hass, _LOGGER, name=f"{DOMAIN}-{config_entry.unique_id}")
+
+    async def drop_message_received(
+        self, topic: str, payload: str, qos: int, retain: bool
+    ) -> None:
+        """Process a received MQTT message."""
+        topic_root = self.config_entry.data[CONF_DATA_TOPIC].removesuffix("/#")
+        try:
+            json_data = json_loads(payload)
+            if topic.startswith(topic_root):
+                structure_key = topic.removeprefix(topic_root).removeprefix("/")
+                _LOGGER.debug(
+                    "New data for %s/%s [%s]: %s",
+                    self.config_entry.data[CONF_HUB_ID],
+                    self.config_entry.data[CONF_DEVICE_ID],
+                    structure_key,
+                    payload,
+                )
+
+                # Create empty dictionary for this structure key if it does not already exist.
+                if structure_key not in self._device_information:
+                    self._device_information[structure_key] = {}
+
+                # Merge incoming data into the existing dictionary.
+                self._device_information[structure_key].update(json_data)
+                self.async_set_updated_data(None)
+        except JSON_DECODE_EXCEPTIONS:
+            _LOGGER.error("Invalid JSON (%s): %s", topic, payload)
+
+    # Device properties
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return a device description."""
+        device_info = DeviceInfo(
+            manufacturer=self.manufacturer,
+            model=self.model,
+            name=self.device_name,
+        )
+        if self.config_entry.data[CONF_DEVICE_TYPE] == DEV_HUB:
+            device_info.update(
+                {"identifiers": {(DOMAIN, self.config_entry.data[CONF_HUB_ID])}}
+            )
+        else:
+            device_info.update(
+                {
+                    "identifiers": {(DOMAIN, self.config_entry.unique_id or "")},
```

Comment:
Don't the other devices have their own identifier?

Reply:
If config_entry.unique_id is always there, use 

`assert config_entry.unique_id` before this so you don't have to do the `or ""`

Reply:
Like, I have the feeling that the moment it ever encounters the `or ""` a lot of stuff will be unexpected, so I'd rather raise an AssertionError than register devices with `""` as identifier. What error do you get?

Reply:
The problem adding the assert was simply that I hadn't set `unique_id` values in my unit test fixtures.  That oversight plus replacing the `or ""` with an assert is fixed in e6bf06244dc9c4728cc63237060c5cca6039a9ce.